### DISCUSSION TITLE
How to log in using Apple Sign In (AuthenticationServices framework)

### DISCUSSION THREAD

**Discussion by Vinzius** [supabase/supabase#1882] [Questions]
Hey,

I tried getting some answers from the Slack community, but without success.

I don't know Apple Sign In / JWT well enough, so my guess is I either 1/ missed something 2/ misconfigured something.

I try to use the native Apple Sign In experience with the AuthenticationServices framework.
+ I followed various tutorials to get the client id / secret, and put them in the supabase settings.
+ I configured the callback endpoint in the app dev account.

When I log-in and get the callback `func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization)`, I'm not sure what to do after with the JWT.
+ I tried to use it directly with supabase, but it tells my the signature is not good.
+ I see no logs in the supabase console, which make sense as the /authorize route has never been called (and Apple isn't aware of it I guess).

Is there a route to call in the API rest?

Thank :)

**Comment by awalias** [comment]
for some reason it won't let me transfer this to the https://github.com/supabase/gotrue-swift repo

@satishbabariya idk if you have any ideas here?

**Reply by libovness** [reply]
@awalias Mind if I just post this again in that repo? @Vinzius is a colleague of mine

**Reply by burggraf** [reply]
Apple authentication takes a lot of steps to get working, but I've written a step-by-step guide on it plus made a video.  Hopefully these are helpful for you.  

Guide -> [Setting up Apple Auth](https://supabase.io/docs/guides/auth/auth-apple)
Video -> [Supabase Authentication: Setting up Apple Auth](https://www.youtube.com/watch?v=6I2JEky20ME)

Please let me know if you have any questions on this.  The difficult part for me was getting the secret key file downloaded from Microsoft turned into a real client secret string.

If you already have your client_id and client_secret working with the Javascript client or using some other method then please ignore this.  I'm not proficient enough in Swift to know if there are differences in using the authentication tokens.

**Reply by Vinzius** [reply]
Hey @burggraf !
It's not an SQL question this time haha =D
The guides that have been shared until now have been really useful to setup the Apple Sign In id/secret!
Our issue is with making them work from an iOS client with Swift code. I think the "gotrue flow" may be different, and is not compatible with existing REST routes (see my other answer below).
Thanks :)

**Reply by burggraf** [reply]
Yeah, @kangmingtay figured out the issue for you and  has already proposed a workaround below.  Hopefully the maintainers of the Swift api repo will be able to add all the new providers soon.  It's hard to keep up with the Supabase team! LOL

**Reply by awalias** [reply]
@libovness does this help? we were missing some providers https://github.com/supabase/gotrue-swift/pull/1

**Reply by libovness** [reply]
@awalias Unfortunately it does not. The current gotrue implementation is great for web authentication, but it doesn't support native Apple Sign In where Apple directly provides the (Apple) JWT to the developer, and not through supabase 

Note that we are happy to help with the Swift implementation as much as we can

cc @Vinzius

**Comment by kangmingtay** [comment]
Hey @Vinzius, I'm not too familiar with Swift myself but I took a look at the [authentication services](https://developer.apple.com/documentation/authenticationservices/implementing_user_authentication_with_sign_in_with_apple) "user authentication sign-in with apple flow" and wanted to suggest a possible workaround:

1. Follow the guide above till the "Handle User Credentials" section. 
2. If the authentication succeeds, when the authentication controller invokes the `authorizationController(controller:didCompleteWithAuthorization:)` delegate function, you can add a call to gotrue's /signup endpoint to register the new user. (This endpoint would sign up a user in our auth schema via an email and password) 
3. If it's an `appleIDCredential`, you should generate a secure fake password and pass it into the /signup endpoint. Else if its a `passwordCredential` sign up with the email and password pair as per-normal. 
4. If the authentication fails, the `authorizationController(controller:didCompleteWithError:)` delegate function is invoked and you can add the necessary flow to handle the failure. 

Hope it helps! :)

**Reply by Vinzius** [reply]
Hey @kangmingtay 

Thanks for taking the time to answer!
I understand your suggestion, but it means it would be an email login, and not an Apple Sign In. And so trying to login from the web would create a different account.

I'm not sure to fully understand the go true flow (while I tried to read a bit the gotrue repository). But from what I understand, the gotrue part on the web takes care of retrieving the JWT from the provider (here Apple), while in the iOS app case, it's done by the application itself.

Would it be theoretically possible to have another API route to send the JWT sent by Apple (from Swift code), validate it, create an account, and return a supabase JWT? Or am I mistaken on some step / how it works?

Cheers :-)

**Reply by kangmingtay** [reply]
Oh ok I see where you're coming from. Yeah you are right that gotrue takes care of retrieving the JWT from the provider and validating it. It's able to validate the JWT because the JWT secret is known to the gotrue server. 

In the IOS app case, it is theoretically possible to have an API route to receive the JWT sent by Apple but gotrue will not be able to validate it because it doesn't know the JWT secret. However, we understand that this is an existing problem for mobile developers using gotrue-auth  (those in the Expo community face this too!) and will be looking to come up with a possible solution in our pipeline. Just hang in tight!

**Reply by catlan** [reply]
I started playing with this a bit: https://github.com/supabase/gotrue-swift/issues/4

**Reply by johndpope** [reply]
any update for ios - @kangmingtay ? it's a bit of deal breaker for migrating stuff across.

**Reply by johndpope** [reply]
fyi - I'm looking at using self hosted apple node js / express / postgress to bypass problems. I've started repo here - it's getting heavier now I need to drop into auth schema to craft a user or dig one out. It would require hosting - nodejs - but if this works - it would solve all the gotrue-auth problems.
For clarity - my hosted nodejs knows the supabase jwt secret - so it can (providing I can craft appropriate dictionary).
https://github.com/johndpope/Sign-in-with-Apple-for-supabase

I've got a bunch of problems
- manually creating / inserting the user with identity provide = apple  (using postgres)
- how to issue the access_token / jwt  
```javascript
	// create an access token
		let claims = {
			"StandardClaims": {
				"sub": newUser.id,
				"aud": "",
				"exp": Math.floor(Date.now() / 1000),
			},
			"Email": newUser.Email,
			"AppMetaData": newUser.AppMetaData,
			"UserMetaData": newUser.UserMetaData,
		}
		console.log("✅ claims:", claims);
		const jwt = sign(claims, config.supabase.jwtSecret);
		console.log("jwt:", jwt);
```

BACKGROuND - this is the user flow supabase needs to implement for native app apple sign in ( not the web flow)
https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api
